pkg/messaging/mqtt/pubsub.go

```diff
@@ -46,31 +47,34 @@ type pubsub struct {
 	subscriptions map[string]subscription
 }
 
+// func (ps *pubsub)
+
 // NewPubSub returns MQTT message publisher/subscriber.
 func NewPubSub(url, queue string, timeout time.Duration, logger log.Logger) (messaging.PubSub, error) {
 	client, err := newClient(url, "mqtt-publisher", timeout)
 	if err != nil {
 		return nil, err
 	}
-	ret := pubsub{
+	ret := &pubsub{
 		publisher: publisher{
 			client:  client,
 			timeout: timeout,
 		},
 		address:       url,
 		timeout:       timeout,
 		logger:        logger,
+		mu:            &sync.RWMutex{},
```

Comment:
No need to use a pointer here.

Reply:
@dborovcanin This is a problem, that I'm still trying to fix.
If I don't initialize sync.RWMutex{} here, then I get the `nil memory reference error` for some reason. 
This doesn't happen for rabbitmq or nats. This is the only fix I found for the problem, I'm still working on it

Reply:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4764be]

goroutine 139 [running]:
testing.tRunner.func1.2({0x93f400, 0xdd7bd0})
        /usr/lib/go-1.18/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
        /usr/lib/go-1.18/src/testing/testing.go:1392 +0x39f
panic({0x93f400, 0xdd7bd0})
        /usr/lib/go-1.18/src/runtime/panic.go:838 +0x207
sync.(*RWMutex).Lock(0x0?)
        /usr/lib/go-1.18/src/sync/rwmutex.go:139 +0x1e
github.com/mainflux/mainflux/pkg/messaging/mqtt.(*pubsub).Subscribe(0xc00007c2d0, {0x9e6f4d, 0x24}, {0xc000029d30, 0xe}, {0xaaf9c8, 0xc000029d28})
        /home/aryan/mainflux/pkg/messaging/mqtt/pubsub.go:77 +0xbc
github.com/mainflux/mainflux/pkg/messaging/mqtt_test.TestPublisher(0xc0002b8ea0?)
        /home/aryan/mainflux/pkg/messaging/mqtt/pubsub_test.go:31 +0x10a
testing.tRunner(0xc0002b8ea0, 0xa1acb8)
        /usr/lib/go-1.18/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
        /usr/lib/go-1.18/src/testing/testing.go:1486 +0x35f
FAIL    github.com/mainflux/mainflux/pkg/messaging/mqtt 12.346s
FAIL
```

Reply:
If you don't use a pointer - you don't have nil. :)